### ISSUE TITLE
snapshot/plot browsing repeatedly knits all Rmd files

### ISSUE THREAD

**Issue by ras44** [rstudio/rstudio#2859]
### System details

    RStudio Edition : Server
    RStudio Version : 1.1.414
    OS Version      :  CentOS 7
    R Version       : 3.4.4

### Steps to reproduce the problem
Install packrat and create a project with packrat enabled.
Install many packages (30 for example).
Create many .Rmd files (about 10).
When you snapshot the library, R knits each .Rmd file multiple times (it seems to be once per package).

### Describe the problem in detail
after installing 32 libraries, I run
```
packrat::snapshot(ignore.stale = TRUE, infer.dependencies = FALSE)
```

Output indicates that it fetches all sources, but then I notice that the 'Packages' window is repeatedly refreshed before the command completes.  If I switch to the 'Files' window, I can see that each .Rmd file in the project is being knitted repeatedly (perhaps knitted once for each library?).  This leads to what seems like a very long duration for the snapshot operation.

In a related issue, I notice that when I have packrat installed and multiple .Rmd files and want to browse plots (by clicking back and forward on the plot navigation icons), the .Rmd files in the project are knitted.  This seems unnecessary and inefficient.

### Describe the behavior you expected
If knitting the files is necessary, then knit them all just once.  Or provide an option to manually override this functionality and not knit all .Rmd files when snapshotting or changing plot windows.

**Comment by kevinushey** [comment]
I believe this is because, as part of the attempt to discover dependencies, Packrat will render R Markdown files (either with a special `evaluate()` hook, or by calling `knitr::knit()`). This is effectively done to extract the R code from an R Markdown document so that dependency discover can be run on that file.

We could fix this in Packrat by moving to a more ad-hoc dependency discovery model -- either implement our own R-code-extractor and then examine that R code, or even just a simple 'grep' for calls to `library()` or `foo::` and such.

**Comment by kevinushey** [comment]
And finally, we could (should?) make this a Packrat project option controlling how dependencies are discovered in R Markdown documents.

**Comment by ras44** [comment]
Yes, a project option would be great.  We tend to use Packrat strictly on an update-as-you-go workflow (no automatic snapshotting or dependency discovery required) and the overhead of the current dependency discovery model significantly impacts iterative workflows.

**Comment by DaveRGP** [comment]
I've found the number of`.Rmd` required to be far less, I only have 4 in a project now, and it was an issue from 2 files onwards.

**Comment by ras44** [comment]
I agree with @DaveRGP- the effects are notable even with very few .Rmd files.  @kevinushey if you could point me in the right direction re the `evaluate()` hooks and where this is managed in the rstudio code, I could try to take a stab at a PR to get the ball moving.

**Comment by DaveRGP** [comment]
Same, I'm happy to help if I can :)

**Comment by kevinushey** [comment]
The associated Packrat code lives here:

https://github.com/rstudio/packrat/blob/215a30239806dfe4837047c6d1deb2b3f7e8285a/R/dependencies.R#L571-L623

https://github.com/rstudio/packrat/blob/215a30239806dfe4837047c6d1deb2b3f7e8285a/R/dependencies.R#L209-L326

I think the fix would need to be part of Packrat rather than RStudio, for what it's worth.

**Comment by ras44** [comment]
@kevinushey would you be open to a `fileDependencies.Rmd.evaluate` option that is TRUE by default, but can be set to FALSE in order to skip the recursive parsing of Rmd files?  I'd be happy to submit a PR for this.  Thanks.

**Comment by kevinushey** [comment]
I think that's reasonable. But maybe the current scheme used by Packrat is flawed overall given these issues and we should change the behavior uniformly.

**Comment by ras44** [comment]
agreed- would you like me to submit the PR as an easy intermediate patch?  I'm not sure if changing the behavior uniformly would imply breaking the 0.4 minor version, and hence a longer ways out for implementation, but I would guess that it would be a bigger undertaking.

**Comment by kevinushey** [comment]
I would be in favor of a PR.
